platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/BindingsBackedCodec.kt

```diff
@@ -33,6 +33,12 @@ import org.gradle.internal.serialize.Serializer
 import kotlin.reflect.KClass
 
 
+/**
+ * An implementation of the Codec protocol that (based on a [Binding.tag]) chooses and delegates
```

Comment:
The binding (a tagged codec) is chosen based on the availability of a `Binding.encoding` for the value being encoded. This is basically implemented as a predicate dispatching on the value type, first available `Binding.encoding` wins and its `Binding.tag` is recorded in the output stream so decoding can be implemented via a fast array lookup.